### PR TITLE
[fx] fix type promotion in `binary_magic_impl`

### PR DESCRIPTION
**Pull request by nkaretnikov**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #91376
* #93143
* #93142
* #93185

### CODE REVIEW

**Comment by nkaretnikov** [comment]
This makes many changes, I know, but I've tried hard to be compatible with Python where possible. It's not always easy since we only have `SymFloat` and `SymInt` while many ops can return `bool` and `complex`. In some cases, you also cannot derive the output type just from input types (see the comments for examples).

**Comment by nkaretnikov** [comment]
@bdhirsh IIUC, Ed might not be available for a bit, so PTAL if you feel comfortable reviewing this. Otherwise, we can just wait for Ed.

File: test/test_dynamic_shapes.py [hunk 1]
```diff
@@ -438,34 +436,6 @@ def forward(self, a_1: f32[s0, s1], b_1: f32[s2, s1]):
...
-    print("expected_failure_sym_magic_methods = {")
-    print("\n".join(out))
-    print("}")
-
-if COLLECT_EXPECT:
-    atexit.register(print_seen)
-
-expected_failure_sym_magic_methods = {
-}
-
```
Comment (nkaretnikov) [inline test/test_dynamic_shapes.py:468]: I don't think there's any reason to keep this around since the test logic is now different, so I removed the whole thing.

File: test/test_dynamic_shapes.py [hunk 2]
```diff
@@ -569,18 +545,30 @@ def test_method(self, fn, first_type, second_type):
...
+        # Only floats here since these will be converted to int if necessary.
+        # We also ignore complex and bool.
+        values = (
+            0.0,
+            1.0,
+            2.5,
+        )
 
-        shape_env = ShapeEnv()
+        neg_values = tuple(-x for x in values)
```
Comment (nkaretnikov) [inline test/test_dynamic_shapes.py:556]: Added more inputs since these create interesting corner cases: short-circuit behavior in some ops, complex output, etc.

File: torch/fx/experimental/symbolic_shapes.py [hunk 3]
```diff
@@ -388,13 +424,27 @@ def binary_magic_impl(self, other):
             raise
         out = sympy.expand(out)
         pytype: Type
+        # This is not strictly correct. In Python, a**b may return complex when
+        # a < 0 and b is a float: (-1)**2.1. Same for sympy.sqrt(-3.14). This
+        # returns a float while both arguments are ints: 2**(-1). Also, max and
+        # min do not type promote. To avoid having data-dependent control flow
+        # here, we just set the type to float if one of the args is a float. In
+        # case of a type mismatch, we assume that it will be detected during
+        # evaluation.
```
Comment (nkaretnikov) [inline torch/fx/experimental/symbolic_shapes.py:433]: See the comments here on various corner-cases related to type promotion.

File: torch/fx/experimental/symbolic_shapes.py [hunk 4]
```diff
@@ -220,7 +226,15 @@ def int_(self):
     def guard_int(self, file, line):
         # TODO: use the file/line for some useful diagnostic on why a
         # guard occurred
-        return int(self.shape_env.evaluate_expr(self.expr))
+        # Because there is no SymBool, we wrap bools into SymInt during
+        # construction. So we have to handle bools here.
```
Comment (nkaretnikov) [inline torch/fx/experimental/symbolic_shapes.py:230]: Maybe mention `wrap_node` here.

File: torch/fx/experimental/symbolic_shapes.py [hunk 5]
```diff
@@ -199,7 +205,15 @@ def int_(self):
...
+        res = self.shape_env.evaluate_expr(self.expr)
+        if res is sympy.sympify(False):
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:211]: ugh lol this sucks (no action needed). Really need to put in SymBool asap

File: torch/fx/experimental/symbolic_shapes.py [hunk 6]
```diff
@@ -211,6 +225,28 @@ def bool_(self):
...
+    # https://github.com/pytorch/pytorch/issues/90900
+    class Pow(sympy.Function):
+        @classmethod
+        def eval(cls, base, exp):
+            if exp == 0:
+                return sympy.Integer(1)
+            elif base == 0 and exp < 0:
+                raise ZeroDivisionError(f"{base} cannot be raised to a negative power")
+            else:
+                return base ** exp
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:238]: bleh, losing the printing here sucks

File: torch/fx/experimental/symbolic_shapes.py [hunk 7]
```diff
@@ -374,13 +410,27 @@ def binary_magic_impl(self, other):
...
+        # min do not type promote. To avoid having data-dependent control flow
+        # here, we just set the type to float if one of the args is a float. In
+        # case of a type mismatch, we assume that it will be detected during
+        # evaluation.
         if method in always_float_magic_methods:
             pytype = float
+        elif method in ("min", "max"):
+            # This doesn't type-promote in Python, but it's data-dependent, so
+            # we always assume float.
+            pytype = float
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:425]: Sorry, so if it doesn't type promote, why is it data dependent?

File: torch/fx/experimental/symbolic_shapes.py (hunk 6 above)
Comment (ezyang) [reply torch/fx/experimental/symbolic_shapes.py:238]: cc @Chillee 

I think I'd be willing to "lie" and have these print as `**` and `/`

**Review by ezyang — APPROVED** [review]
@albanD you may want to take a look at the test changes

**Review by albanD — APPROVED** [review]
Nice improvement! Thanks!

File: torch/fx/experimental/symbolic_shapes.py (hunk 7 above)
Comment (nkaretnikov) [reply torch/fx/experimental/symbolic_shapes.py:425]: You need to eval to be able to compare, so cannot just rely on the types.

File: test/test_dynamic_shapes.py [hunk 8]
```diff
@@ -545,7 +515,9 @@ def guard_fn(v):
                 out = lambda_apply(sym_inp1)
             else:
                 out = lambda_apply(sym_inp1, inp2)
-            self.assertEqual(guard_fn(out), ref_out)
+            if out is not None:
```
[the remark this answers is not in the document]
Comment (nkaretnikov) [reply test/test_dynamic_shapes.py]: Nice catch! I think the intention was to check when a ref and impl both raise an exception, but the tests pass w/o that, so I've just removed this. Thanks!

File: torch/fx/experimental/symbolic_shapes.py (hunk 6 above)
Comment (nkaretnikov) [reply torch/fx/experimental/symbolic_shapes.py:238]: When is printing an issue for these? These fully eval. Are you proposing not to eval the `else` cases and do the same as in `FloorDiv`?

**Review by albanD — APPROVED** [review]

File: torch/fx/experimental/symbolic_shapes.py (hunk 6 above)
Comment (nkaretnikov) [reply torch/fx/experimental/symbolic_shapes.py:238]: I don't think so. `truediv` is just this: `lambda a, b: TrueDiv(a, b)`, right? And `TrueDiv` is fully eval'd (on our side), so it dispatches to however `/` is handled. Seems fine to me:

```py
[ins] In [5]: TrueDiv(2,3)
Out[5]: 2/3

[ins] In [6]: Pow(2,3)
Out[6]: 8
```

Comment (ezyang) [reply torch/fx/experimental/symbolic_shapes.py:238]: I'm talking about when we print expressions with symbolic variables in them

Comment (nkaretnikov) [reply torch/fx/experimental/symbolic_shapes.py:238]: Those work fine as well. Note that we dispatch to `**` or `/` in `else`, so they take care of printing.

```py
[ins] In [8]: TrueDiv(sympy.Symbol("y"),sympy.Symbol("x"))
Out[8]: y/x

[ins] In [9]: Pow(sympy.Symbol("y"),sympy.Symbol("x"))
Out[9]: y**x
```

**Comment by nkaretnikov** [comment]
Note: I merged when everything was green (including trunk and inductor tests). I did see some failures before, but those were fixed after re-running the tests. So I assumed flaky tests. I'll investigate.

**Comment by malfet** [comment]
@nikitaved just to be clear, I'm not sure your PR have caused, but but right now there seem to be a correlation on HUD, see https://hud.pytorch.org/hud/pytorch/pytorch/master/1?per_page=50&name_filter=inductor_timm%2C%201%2C%202
Also, reverting base-PR as well, will reland if it would not fix the issue

**Comment by ZainRizvi** [comment]
Also, adding the `ciflow/inductor` label to your PR will help trigger extra inductor tests which normally don't get run on PRs but do get run once they're merged into master (and it was one of those which failed)

**Comment by nkaretnikov** [comment]
Ah, maybe this one landed without the inductor label, hmm. Okay, let's try reopening it and see if it repros.

File: torch/fx/experimental/symbolic_shapes.py [hunk 9]
```diff
@@ -243,6 +255,28 @@ def bool_(self):
...
+
+    # Overloaded to be compatible with regular Python.
+    # https://github.com/pytorch/pytorch/issues/90900
+    class TrueDiv(sympy.Function):
+        @classmethod
+        def eval(cls, base, divisor):
+            if divisor.is_zero:
+                raise ZeroDivisionError("division by zero")
+            else:
+                return base / divisor
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:278]: Probably need to manually implement the assumptions stuff here too

Comment (nkaretnikov) [reply torch/fx/experimental/symbolic_shapes.py:278]: I don't think so. These never return None, so they are replaced with the underlying sympy exprs, eg:

```
[ins] In [3]: Pow(sympy.Symbol("y"),sympy.Symbol("x"))
Out[3]: y**x
```

File: test/test_dynamic_shapes.py [hunk 10]
```diff
@@ -15,8 +15,9 @@
 from torch.utils._pytree import tree_map
 from torch.fx.experimental import symbolic_shapes
 from torch.fx.experimental.proxy_tensor import make_fx
-from torch.fx.experimental.symbolic_shapes import FloorDiv, ShapeEnv, \
-    guard_bool, guard_int, guard_float, SymNode, sym_sqrt, sym_int, sym_float, to_node
+from torch.fx.experimental.symbolic_shapes import SymNode, \
+    sym_sqrt, sym_int, sym_float to_node, GuardOnDataDependentSymNode, \
```
Comment (nkaretnikov) [inline test/test_dynamic_shapes.py:19]: missing comma after sym_float, will fix
